### PR TITLE
[Add] add example_gqa_bwd

### PR DESCRIPTION
**Pull request by wch0810**
## Description
Migrate GQA backward kernel, implemented in `example_gqa_bwd.py` based on TileLang + AscendC.
Add design document for implementation.

Tested on Ascend NPU, fp16, B=8 H=32 H_kv=2 groups=16. Precision verified against PyTorch golden (atol=1e-2, rtol=1e-2).

## Non-causal (D_qk=192, D_v=128 -- golden config)
| Kernel                      | Q 序列长度 | K/V 序列长度 | 延迟 (ms) | 算力 (TFlops) | 精度 (max_diff) |
| --------------------------- | ---------- | ------------ | --------- | ------------- | --------------- |
| TileLang Forward v1         | 1024       | 1024         | 14.61     | 11.76         | 2.44e-04        |
| TileLang Forward v4         | 1024       | 1024         | 10.10     | 17.01         | 2.44e-04        |
| TileLang Backward (pipeline)| 1024       | 1024         | 34.44     | 12.97         | 9.77e-04        |
| TileLang Fwd(v4)+Bwd        | 1024       | 1024         | 44.55     | 13.88         | —               |
| PyTorch Forward only        | 1024       | 1024         | 7.30      | 23.54         | —               |
| PyTorch Fwd+Bwd (e2e)       | 1024       | 1024         | 20.02     | 30.89         | —               |

- v4 vs v1 forward: 1.45x speedup
- v4 forward vs PyTorch forward: 0.72x

## Causal (D_qk=128, D_v=128)
| Kernel                      | Q 序列长度 | K/V 序列长度 | 延迟 (ms) | 算力 (TFlops) | 精度 (max_diff) |
| --------------------------- | ---------- | ------------ | --------- | ------------- | --------------- |
| TileLang Forward v1         | 1024       | 1024         | 10.17     | 6.76          | 1.95e-03        |
| TileLang Forward v4         | 1024       | 1024         | 9.45      | 7.27          | 1.95e-03        |
| TileLang Backward (pipeline)| 1024       | 1024         | 22.53     | 7.62          | 7.81e-03        |
| TileLang Fwd(v4)+Bwd        | 1024       | 1024         | 31.98     | 7.52          | —               |
| PyTorch Forward only        | 1024       | 1024         | 10.57     | 6.50          | —               |
| PyTorch Fwd+Bwd (e2e)       | 1024       | 1024         | 24.72     | 9.73          | —               |

- v4 forward vs PyTorch forward: 1.12x speedup (causal mode beats PyTorch)

## Notes
- Forward v4: L0 double buffering + batched softmax + Fixed Core + causal loop range optimization (num_stages=8).
- Backward: pipeline kernel with fine-grained set_flag/wait_flag sync in C scope (num_stages=8).
- v1 is skipped in causal + D_qk > 128 due to L0C overflow (known architecture limitation).
- End-to-end ~2.4x gap vs PyTorch is an architecture-level limitation (C/V physical separation forces workspace GM round-trip).

## Checklist
- [x] GQA backward kernel implementation
- [x] Basic function verification passed

### CODE REVIEW

**Review by LLMZhangYC — APPROVED** [review]
lgtm

**Review by ChaoyangJi — APPROVED** [review]
approve

**Comment by dlaerbinjwy** [comment]
@wch0810 
我们在运行全量CI时发现该action报错了，核查后发现该错误是一个偶发错误，其中报错的算子是你本PR提交的
https://github.com/tile-ai/tilelang-ascend/actions/runs/31132906264

[FAILED] ../examples_experiment/example_gqa_bwd/test_gqa_bwd.py (Exit: 1)
  Last line:  (function npuSynchronizeDevice)
  EE9999: Inner Error!
  EE9999[PID: 3707517] 2026-08-07-00:35:49.457.978 (EE9999):
    rtDeviceSynchronizeWithTimeout execution failed, reason=aicore exception
    wait for compute device to finish failed, runtime result = 507015.

我们注意到507015 这个码在你 #1486 的 commit 里出现过(9275b915 "fix perf aicore-507015 exception")，想请你是不是同一个问题复发，能不能请你这边核查一下，如果有任何的消息进展麻烦你通知我一下，谢谢~
